Add all callback-related args to all widgets

vdonato opened the pull request:
Note that we don't actually do anything with these args aside from
verify that the widget is well-behaved and follows some rules we have
around callbacks. Actually hooking them in can happen once session and
widget state are more properly unified, and we might as well get this
routine work out of the way now.

I hesitantly but intentionally omitted a good amount of type info from
both the docstrings and function signatures. This feels a bit weird, but
I think it's necessary because

  * when sphinx/RTD renders documentation, it uses the function
    signature to do so, and unfortunately things look pretty awful for
    long function signatures with complex types. See the screenshot
    below.

<img width="699" alt="Screen Shot 2021-06-04 at 4 29 16 PM" src="https://user-images.githubusercontent.com/3144420/120871946-0bbf2b00-c552-11eb-94c5-8eef4e08f1c4.png">

  * existing documentation in the API Reference does not state that some
    args are optional in the type but does do so in the following
    explanatory text, and I didn't want to take the time to also rewrite
    parts of the docstrings for every widget.

We can probably look to resolve these issues later, but I think doing so
is outside of the scope of this work on session state.

vdonato committed:
Add all callback-related args to all widgets

Note that we don't actually do anything with these args aside from
verify that the widget is well-behaved and follows some rules we have
around callbacks. Actually hooking them in can happen once session and
widget state are more properly unified, and we might as well get this
routine work out of the way now.

I hesitantly but intentionally omitted a good amount of type info from
both the docstrings and function signatures. This feels a bit weird, but
I think it's necessary because

  * when sphinx/RTD renders documentation, it uses the function
    signature to do so, and unfortunately things look pretty awful for
    long function signatures with complex types.
  * existing documentation in the API Reference does not state that some
    args are optional in the type but does do so in the following
    explanatory text, and I didn't want to take the time to also rewrite
    parts of the docstrings for every widget.

We can probably look to resolve these issues later, but I think doing so
is outside of the scope of this work on session state.

kantuni commented on lib/streamlit/elements/button.py:
```diff
@@ -39,31 +39,31 @@ class ButtonMixin:
         label : str
             A short label explaining to the user what this button is for.
-        key : Optional[str]
+        key : str
```
Should we preserve the *Optional* part here?

vdonato replied:
I intentionally removed it to stay consistent with the rest of our documentation, where we don't type our optional fields so accurately and instead just mention that the parameter is optional in its description.

I definitely think there's some merit to making the types here more correct, but it's something that'd probably be best done outside of the session state feature branch since it'll require us to change docstrings for most of our API.

kantuni commented on lib/streamlit/elements/file_uploader.py:
```diff
@@ -58,6 +67,16 @@ def file_uploader(
             A tooltip that gets displayed next to the file uploader.
 
+        on_change : callable
+            An optional callback invoked when this file_uploader's value
```
`file_uploader's` → `file uploader's` (just to be consistent)

kantuni commented on lib/streamlit/elements/color_picker.py:
```diff
@@ -39,7 +49,14 @@ def color_picker(self, label, value=None, key=None, help=None):
-            A tooltip that gets displayed next to the color picker.
+            An optional tooltip that gets displayed next to the color picker.
+        on_change : callable
+            An optional callback invoked when this color picker's value
```
`color_picker's` (just to be consistent with other docstrings)

vdonato replied:
Done

kantuni commented on lib/tests/streamlit/delta_generator_test.py:
```diff
@@ -138,43 +139,6 @@ def test_sidebar_nonexistent_method(self):
             "Did you mean `st.echo()`?",
         )
 
-    @parameterized.expand(
```
Why were these tests removed?

vdonato replied:
I might be missing something, but I didn't see the value in a test that only asserts that the function signature of a few select widgets is fixed. It seemed a bit too brittle / not useful thing to validate to me, so I removed it when the test started failing due to the `st.checkbox` signature changing.

Happy to add the tests back with an explanatory comment if there's a good reason these were here.

vdonato committed:
Change "color picker" -> color_picker in docstring

vdonato committed:
Readd is_new_value session_state method

We readded the method under a slightly different name is_new_state_value.
This was removed in a PR probably because it was unused/seemed like dead
code, but we need it to enforce some session_state API rules.

vdonato commented:
Merging despite the cypress test failures since those already exist on the `feature/session-state` branch (we knowingly merged a PR even though not all cypress tests pass because we're prioritizing getting a wheel file out)
